### PR TITLE
 #161342852 Add Search Functionality and Custom filtering

### PR DESCRIPTION
**Pull request by ucheg6**
#### What does this PR do?
- Add Search Functionality and Custom filtering based on parameters
#### Description of Task to be completed?
- create a migration file to alter articles tables with a column searchVectors that contains ts_vectors dataType
- create an index and a trigger in the articles table to collate info from title, body and description
- create a function in the article controller that takes in searchTerm, author or tag parameters
- filter search based on the availability of the above parameters
#### How should this be manually tested?
- Git clone the repository
- cd into the `haven-ah-backend` folder
- run `npm run start-dev`
- Fire up the route `localhost:5000/api/v1/articles/search?:queries` as a post request
- To test the entire application, run `npm run test`
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[ #161342852](https://www.pivotaltracker.com/story/show/161342852)
#### Screenshots (if appropriate)
screen shot 2018-11-14 at 1 58 28 pm
screen shot 2018-11-14 at 1 59 06 pm
screen shot 2018-11-14 at 2 00 31 pm

com/34472497/48484304-13911580-e816-11e8-914d-27331394dd22.png">
screen shot 2018-11-14 at 1 59 25 pm
screen shot 2018-11-14 at 1 59 42 pm

#### Questions:
N/A

### CODE REVIEW

**Review by victor-nweze-deel — COMMENTED** [review]
Nice work @ucheg6

File: server/repository/articleRepository.js [hunk 1]
```diff
@@ -110,6 +112,86 @@ class ArticleRepository {
...
+      const Author = await User.findByPk(result.userid, {
+        attributes: ['firstName', 'lastName', 'username', 'imageUrl']
+      });
+      result.Author = Author;
+    }));
+
+
+    return keywordsResults;
+  }
 }
```
Comment (TheoOkafor) [inline server/repository/articleRepository.js:251]: Nice work here @ucheg6 👍

**Review by TheoOkafor — COMMENTED** [review]
Great job @ucheg6

File: server/controllers/article.js [hunk 2]
```diff
@@ -150,6 +150,25 @@ class Article {
...
+
+  /**
+   * Search for an article
+   * @param {object} request Request Object
+   * @param {object} response Response Object
+   * @returns {object} searched articles
+   */
+  static async searchArticle(request, response) {
+    const { author, tag, keywords } = request.query;
+    const foundArticle = await articleRepo.search(keywords, author, tag);
```
Comment (seyi-adeleke) [inline server/controllers/article.js]: are you returning only one article?

File: server/controllers/article.js [hunk 3]
```diff
@@ -150,6 +150,25 @@ class Article {
...
+    if (foundArticle.length === 0) {
```
Comment (seyi-adeleke) [inline server/controllers/article.js]: `!foundArticle.length  `

File: server/middlewares/validations.js [hunk 4]
```diff
@@ -235,6 +235,29 @@ class inputValidator {
...
   }
+
+  /**
+   * @description searchValidator
+   * @param {object} request http request object
+   * @param {object} response http response object
+   * @param {object} next callback function
+   * @returns {object} http response object
+   */
+  static searchValidator(request, response, next) {
```
Comment (seyi-adeleke) [inline server/middlewares/validations.js:246]: shouldnt you validate that author, tag, and keywords are strings?

**Review by seyi-adeleke — CHANGES_REQUESTED** [review]

File: server/controllers/article.js (hunk 2 above)
Comment (ucheg6) [reply server/controllers/article.js]: Nope, I'll handle this. thanks

File: server/controllers/article.js (hunk 3 above)
Comment (ucheg6) [reply server/controllers/article.js]: This worked Thanks.

File: server/middlewares/validations.js (hunk 4 above)
Comment (ucheg6) [reply server/middlewares/validations.js:246]: This has been done

**Review by seyi-adeleke — APPROVED** [review]

**Pull request merged**